geomstats/geometry/base.py

```diff
@@ -699,3 +697,195 @@ def projection(self, point):
         """
+
+
+class ImmersedSet(Manifold, abc.ABC):
+    """Class for manifolds embedded in a vector space by an immersion.
+
+    The manifold is represented with intrinsic coordinates, such that
+    the immersion gives a parameterization of the manifold in these
+    coordinates.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the embedded manifold.
+    """
+
+    def __init__(self, dim, equip=True):
+        super().__init__(
+            dim=dim, shape=(dim,), default_coords_type="intrinsic", equip=equip
+        )
+        self.embedding_space = self._define_embedding_space()
+
+    @staticmethod
+    def default_metric():
+        """Metric to equip the space with if equip is True."""
+        return PullbackMetric
+
+    @abc.abstractmethod
+    def _define_embedding_space(self):
+        """Define embedding space of the manifold.
+
+        Returns
+        -------
+        embedding_space : Manifold
+            Instance of Manifold.
+        """
+
+    @abc.abstractmethod
+    def immersion(self, point):
+        """Evaluate the immersion function at a point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., dim]
+            Point in the immersed manifold.
+
+        Returns
+        -------
+        immersion : array-like, shape=[..., dim_embedding]
+            Immersion of the point.
+        """
+
+    def tangent_immersion(self, tangent_vec, base_point):
+        """Evaluate the tangent immersion at a tangent vec and point.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., dim]
+        base_point : array-like, shape=[..., dim]
+            Point in the immersed manifold.
+
+        Returns
+        -------
+        tangent_vec_emb : array-like, shape=[..., dim_embedding]
+        """
+        jacobian_immersion = self.jacobian_immersion(base_point)
+        return gs.matvec(jacobian_immersion, tangent_vec)
+
+    def jacobian_immersion(self, base_point):
+        """Evaluate the Jacobian of the immersion at a point.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., dim]
+            Point in the immersed manifold.
+
+        Returns
+        -------
+        jacobian_immersion : array-like, shape=[..., dim_embedding, dim]
+        """
+        return gs.autodiff.jacobian_vec(self.immersion)(base_point)
+
+    def hessian_immersion(self, base_point):
+        """Compute the Hessian of the immersion.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., dim]
+            Base point.
+
+        Returns
+        -------
+        hessian_immersion : array-like, shape=[..., embedding_dim, dim, dim]
+            Hessian at the base point
+        """
+        return gs.autodiff.hessian_vec(
+            self.immersion, func_out_ndim=self.embedding_space.dim
+        )(base_point)
+
+    def is_tangent(self, vector, base_point, atol=gs.atol):
```

Comment:
These are inherited from `Manifold`, why override them here and allow not to implement them?

Reply:
This is still an underdeveloped class. The main idea of allowing not implementing it is that for using the `PullbackMetric`, these methods are kind of unnecessary (see for example notebook 9). Later we should probably remove this, but I think for now this approach is convenient.